File: 22-generate-parentheses/generate-parentheses.go
func generateParenthesis(n int) []string {
    result := []string{}

    var backtrack func(current string, open, close int)
    backtrack = func(current string, open, close int) {
        if len(current) == n*2 {
            result = append(result, current)
            return
        }

        if open < n {
            backtrack(current+"(", open+1, close)
        }

        if close < open {
            backtrack(current+")", open, close+1)
        }
    }

    backtrack("", 0, 0)
    return result
}

File: 22-generate-parentheses/generate-parentheses.rs
impl Solution {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut result = Vec::new();

        fn backtrack(result: &mut Vec<String>, current: String, open: i32, close: i32, max: i32) {
            if current.len() == (max * 2) as usize {
                result.push(current);
                return;
            }

            if open < max {
                backtrack(result, current.clone() + "(", open + 1, close, max);
            }

            if close < open {
                backtrack(result, current.clone() + ")", open, close + 1, max);
            }
        }

        backtrack(&mut result, String::new(), 0, 0, n);
        result
    }
}
